qiskit/providers/aer/aer_snapshots.py

```diff
@@ -0,0 +1,332 @@
+        tuples = []
+        if isinstance(self, QuantumCircuit):
+            for register in self.qregs:
+                tuples.append(register)
+        if not tuples:
+            raise ExtensionError('no qubits for snapshot')
+        qubits = []
+        for tuple_element in tuples:
+            if isinstance(tuple_element, QuantumRegister):
+                for j in range(tuple_element.size):
+                    qubits.append(tuple_element[j])
+            else:
+                qubits.append(tuple_element)
+    return self.append(
+        SnapshotStatevector(
+            label,
+            snapshot_type=snapshot_type,
+            num_qubits=len(qubits),
+            params=params), qubits)
+
+QuantumCircuit.snapshot_statevector = snapshot_statevector
+
+
+class SnapshotStabilizer(Snapshot):
+    def __init__(self,
+                 label,
+                 snapshot_type='stabilizer',
+                 num_qubits=0,
+                 num_clbits=0,
+                 params=None):
+        super().__init__(label, snapshot_type, num_qubits, num_clbits, params)
+
+def snapshot_stabilizer(self,
+                        label,
+                        snapshot_type='stabilizer',
+                        qubits=None,
+                        params=None):
+
+    # Convert label to string for backwards compatibility
+    if not isinstance(label, str):
+        warnings.warn(
+            "Snapshot label should be a string, "
+            "implicit conversion is deprecated.", DeprecationWarning)
+        label = str(label)
+    # If no qubits are specified we add all qubits so it acts as a barrier
+    # This is needed for full register snapshots like statevector
+    if isinstance(qubits, QuantumRegister):
+        qubits = qubits[:]
+    if not qubits:
+        tuples = []
+        if isinstance(self, QuantumCircuit):
+            for register in self.qregs:
+                tuples.append(register)
+        if not tuples:
+            raise ExtensionError('no qubits for snapshot')
+        qubits = []
+        for tuple_element in tuples:
+            if isinstance(tuple_element, QuantumRegister):
+                for j in range(tuple_element.size):
+                    qubits.append(tuple_element[j])
+            else:
+                qubits.append(tuple_element)
+    return self.append(
+        SnapshotStabilizer(
+            label,
+            snapshot_type=snapshot_type,
+            num_qubits=len(qubits),
+            params=params), qubits)
+
+QuantumCircuit.snapshot_stabilizer = snapshot_stabilizer
+
+class SnaphotDensityMatrix(Snapshot):
+    def __init__(self,
+                 label,
+                 snapshot_type='density_matrix',
+                 num_qubits=0,
+                 num_clbits=0,
+                 params=None):
+        super().__init__(label, num_qubits, num_clbits, params)
+
+def snapshot_density_matrix(self,
+                            label,
+                            snapshot_type='density_matrix',
```

Comment:
Is there a reason the `snapshot_type` is taken as a parameter? It seems like this could allow you to do
`circuit.snapshot_density_matrix('label', snapshot_type='stabilizer') ` which means there are conflicting snapshot types.

Reply:
Good catch. I've fixed this